File: Java/sort/SleepSort.java
/*
    Sleep sort is a sorting algorithm used to sort non-negative integers. Sleep sort works by creating a separate thread for each item in the input which is to be sorted and then each thread sleeps for an amount of time which is proportional to the value of corresponding array element. Therefore, the thread which is having the least amount of sleep time wakes up first and the element gets printed and then the second least element and so on. The largest element wakes up after a long time and then the element gets printed at the last and thus the output obtained is a sorted array.
        
*/

import java.util.Scanner;
import java.util.concurrent.*;

public class SleepSort {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter the number of elements in an array: ");
        int number = sc.nextInt();

        System.out.println("\nEnter the elements in an array: ");
        int array[] = new int[number];

        for (int i = 0; i < number; i++) {
            array[i] = sc.nextInt();

        }
        sleepSort(array);
    }

    public static void sleepSort(int[] array) {
        
        // CountDownLatch in java is used to make sure that a task waits for other threads before it starts.
        // Here the task is going to wait for threads which is equal to array length before it starts.
        final CountDownLatch countDone = new CountDownLatch(array.length);
        System.out.println("\nThe sorted array is: ");
        
        for (final int num : array) {
            // Creates a new thread.
            new Thread(new Runnable() {
                public void run() {
                    // Decrements the count of the countDone
                    countDone.countDown();
                    try {
                        //Causes the current thread to wait until the countDone has counted down to zero
                        countDone.await();
                        Thread.sleep(num);
                        System.out.println(num);

                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }).start();
        }
    }

}

/*
 * Output -
 * 
 * Enter the number of elements in an array: 7
 * 
 * Enter the elements in an array: 10 5 9 77 30 981 14
 * 
 * The sorted array is: 5 9 14 10 30 77 981
 *
 
Time Complexity -
    1. Best case - O(NlogN + max(input))
    2. Average case - O(NlogN + max(input))
    3. Worst case - O(NlogN + max(input))
    where N = number of elements in the input array, and array = input array elements
    
Space Complexity - O(N)
 
 */

File: Python/sort/Sleep_sort.py
'''
  Sleep sort is a sorting algorithm used to sort non-negative integers. Sleep sort works by creating a separate thread for each item in the input which is to be
  sorted and then each thread sleeps for an amount of time which is proportional to the value of corresponding array element. Therefore, the thread which is having
  the least amount of sleep time wakes up first and the element gets printed and then the second least element and so on. The largest element wakes up after a long
  time and then the element gets printed at the last and thus the output obtained is a sorted array.     
'''
  
import threading
import time

lock = threading.Lock()

class SleepSortThread(threading.Thread):
    def __init__(self, val):
        self.val = val
        threading.Thread.__init__(self)

    def run(self):
        global lock
        # Thread is made to sleep in proportion to value
        time.sleep(self.val * 0.01)
        # Acquire lock
        lock.acquire()
        print(self.val, end = " ")
        # Release lock
        lock.release()

def SleepSort(list):
    thread_array = []

    # Intialize a thread corresponding to each element in list
    for i in list:
        thread = SleepSortThread(i)
        thread_array.append(thread)

    # Start all Threads
    for i in thread_array:
        i.start()

    # Wait for all threads to terminate
    for i in thread_array:
        i.join()


if __name__ == "__main__" :
    unsorted_array =[]
    number_of_elements = int(input("Enter the number of elements : "))
    print("Enter the elements: ")
    for i in range(number_of_elements):
        element = int(input())
        unsorted_array.append(element)

    print("The elements in sorted manner are: ")
    sorted_array = SleepSort(unsorted_array)

'''
Enter the number of elements : 7

Enter the elements: 
10
20
40
5
9
88
104

The elements in sorted manner are:
5 9 10 20 40 88 104


Time Complexity -
    1. Best case - O(NlogN + max(input))
    2. Average case - O(NlogN + max(input))
    3. Worst case - O(NlogN + max(input))
    where N = number of elements in the input array, and array = input array elements
    
Space Complexity - O(N)

'''
